File: db/equip.go
// Package db 裝備相關：初始穿搭、裸奔判定、物品種子。對齊裝備分頁規格 §五。
package db

import (
	"encoding/json"

	"singularity_world/store"
)

// StarterEquipment 依性別回傳初始裝備 JSON（裝備分頁規格 §5.1）。
func StarterEquipment(gender string) string {
	if gender == "F" {
		return `{"body":"starter_body_f","legs":"starter_legs_f","feet":"starter_feet_f"}`
	}
	return `{"body":"starter_body_m","legs":"starter_legs_m","feet":"starter_feet_m"}`
}

// IsNaked 檢查 equipment_slots JSON，body 或 legs 任一為空即為「衣不蔽體」（裝備分頁規格 §5.2）。
func IsNaked(equipmentSlots string) bool {
	if equipmentSlots == "" {
		return true
	}
	var slots map[string]string
	if err := json.Unmarshal([]byte(equipmentSlots), &slots); err != nil {
		return true
	}
	return slots["body"] == "" || slots["legs"] == ""
}

// GetItemNames 依 equipment_slots JSON 查物品名稱。
func GetItemNames(equipmentSlots string) (map[string]string, error) {
	result := make(map[string]string)
	if equipmentSlots == "" {
		return result, nil
	}
	var slots map[string]string
	if err := json.Unmarshal([]byte(equipmentSlots), &slots); err != nil {
		return result, nil
	}
	for slot, itemID := range slots {
		if itemID == "" {
			continue
		}
		if store.Default != nil {
			if it := store.Default.GetItem(itemID); it != nil {
				result[slot] = it.Name
			}
		}
	}
	return result, nil
}

// GetItemDescs 依 equipment_slots JSON 查物品描述。
func GetItemDescs(equipmentSlots string) map[string]string {
	result := make(map[string]string)
	if equipmentSlots == "" {
		return result
	}
	var slots map[string]string
	if err := json.Unmarshal([]byte(equipmentSlots), &slots); err != nil {
		return result
	}
	for slot, itemID := range slots {
		if itemID == "" {
			continue
		}
		if store.Default != nil {
			if it := store.Default.GetItem(itemID); it != nil {
				result[slot] = it.Description
			}
		}
	}
	return result
}

// SeedItems 初始化物品種子；現由 data/items.json 於 store.Init 載入，此函式為 no-op。
func SeedItems() error {
	return nil
}

File: src/db/equip.rs
// 初始裝備與裸奔判定，對齊 Go db/equip.go。

use std::collections::HashMap;

use serde_json::Value;

use crate::store;

/// 依性別回傳初始裝備 JSON（對齊 Go `StarterEquipment`）。
#[must_use]
pub fn starter_equipment(gender: &str) -> String {
    if gender == "F" {
        r#"{"body":"starter_body_f","legs":"starter_legs_f","feet":"starter_feet_f"}"#.to_string()
    } else {
        r#"{"body":"starter_body_m","legs":"starter_legs_m","feet":"starter_feet_m"}"#.to_string()
    }
}

/// `body` 或 `legs` 任一為空即「衣不蔽體」（對齊 Go `IsNaked`）。
#[must_use]
pub fn is_naked(equipment_slots: &str) -> bool {
    if equipment_slots.is_empty() {
        return true;
    }
    let Ok(Value::Object(map)) = serde_json::from_str::<Value>(equipment_slots) else {
        return true;
    };
    let body = map.get("body").and_then(|v| v.as_str()).unwrap_or("");
    let legs = map.get("legs").and_then(|v| v.as_str()).unwrap_or("");
    body.is_empty() || legs.is_empty()
}

/// 依 `equipment_slots` 查槽位→物品名稱（對齊 Go `GetItemNames`）。
#[must_use]
pub fn get_item_names(equipment_slots: &str) -> HashMap<String, String> {
    let mut result = HashMap::new();
    if equipment_slots.is_empty() {
        return result;
    }
    let Ok(Value::Object(map)) = serde_json::from_str::<Value>(equipment_slots) else {
        return result;
    };
    let Some(st) = store::get_store() else {
        return result;
    };
    let s = st.read().unwrap();
    for (slot, v) in map {
        let Some(item_id) = v.as_str() else { continue };
        if item_id.is_empty() {
            continue;
        }
        if let Some(it) = s.get_item(item_id) {
            result.insert(slot, it.name.clone());
        }
    }
    result
}

/// 依 `equipment_slots` 查槽位→物品描述（對齊 Go `GetItemDescs`）。
#[must_use]
pub fn get_item_descs(equipment_slots: &str) -> HashMap<String, String> {
    let mut result = HashMap::new();
    if equipment_slots.is_empty() {
        return result;
    }
    let Ok(Value::Object(map)) = serde_json::from_str::<Value>(equipment_slots) else {
        return result;
    };
    let Some(st) = store::get_store() else {
        return result;
    };
    let s = st.read().unwrap();
    for (slot, v) in map {
        let Some(item_id) = v.as_str() else { continue };
        if item_id.is_empty() {
            continue;
        }
        if let Some(it) = s.get_item(item_id) {
            result.insert(slot, it.description.clone());
        }
    }
    result
}

/// 物品種子由 `items.json` 載入；no-op（對齊 Go `SeedItems`）。
pub fn seed_items() -> anyhow::Result<()> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn starter_f_differs_from_m() {
        assert_ne!(starter_equipment("F"), starter_equipment("M"));
        assert!(starter_equipment("M").contains("starter_body_m"));
    }

    #[test]
    fn naked_empty_slots() {
        assert!(is_naked(""));
        assert!(is_naked("{}"));
        assert!(!is_naked(r#"{"body":"x","legs":"y"}"#));
        assert!(is_naked(r#"{"body":"","legs":"y"}"#));
    }
}
